lib/gollum-lib/filter/emoji.rb

```diff
@@ -0,0 +1,23 @@
+# ~*~ encoding: utf-8 ~*~
+
+# Emoji
+#
+# Render emoji such as :smile:
+class Gollum::Filter::Emoji < Gollum::Filter
+  EMOJI_TAG_PATTERN = %r{
+    (?<!\[{2})
+    :(?:([\w-]|</?em>)+):
+    (?!\]{^2})
+  }ix
+
+  def extract(data)
+    data
+  end
+
+  def process(data)
+    data.gsub EMOJI_TAG_PATTERN do |tag|
+      name = tag[1..-2].gsub(%r{</?em>}, '_')   # undo emphasis
```

Comment:
Actually this is why the filter chain is traversed twice (the second time in reverse). You could extract `name` in the `extract` method and replace that string with a temporary placeholder string that is not in danger of being parsed by another filter (the tag filter uses `"TAG#{Digest::SHA1.hexdigest(tag)}TAG"`). Then, in `process` you could replace that placeholder with the HTML string (and the `name` variable should still be available). Take a look at how the tags filter works for an example.